tree_migration_milestone1.js

```diff
@@ -0,0 +1,125 @@
+
+    const count = await knex.select(knex.raw(`count(*) ${base_query_string}`));
+    const query_stream = knex.raw(`select t.* ${base_query_string}`).stream();
+
+    var bar = new ProgressBar("Migrating [:bar] :percent :etas", {
+      width: 20,
+      total: +count[0].count,
+    });
+
+    const attributes_formatter = (attributes) => {
+      if (attributes.length <= 0) return null;
+      const attributes_in_json_format = { entries: [] };
+      for (const { key, value } of attributes) {
+        const new_entry = { key, value };
+        attributes_in_json_format.entries.push(new_entry);
+      }
+    };
+
+    const create_target_objects = (
+      {
+        uuid,
+        time_created,
+        device_identifier,
+        planter_id,
+        planter_photo_url,
+        planter_identifier,
+        gps_accuracy,
+        image_url,
+        lat,
+        lon,
+        note,
+        id,
+        rejection_reason,
+        time_updated,
+      },
+      tree_attributes
+    ) => {
+      const base_target_object = Object.freeze({
+        id: uuid,
+        attributes: attributes_formatter(tree_attributes),
+        capture_taken_at: time_created,
+        created_at: time_created,
+        device_identifier,
+        field_user_id: planter_id,
+        field_username: planter_identifier,
+        lat,
+        lon,
+        updated_at: time_updated,
+      });
+      const raw_capture_to_insert = Object.freeze({
+        ...base_target_object,
+        field_user_photo_url: planter_photo_url,
+        gps_accuracy,
+        image_url,
+        reference_id: id,
+        rejection_reason,
+        status: "approved",
+        note,
+      });
+      return {
+        raw_capture_to_insert,
+        raw_capture_feature_to_insert: base_target_object,
+      };
+    };
+
+    for await (const tree_object of query_stream) {
+      const trx = await knex.transaction();
+      try {
+        const tree_attributes = await trx
+          .select()
+          .table("public.tree_attributes")
+          .where("tree_id", +tree_object.id);
+
+        const { raw_capture_to_insert, raw_capture_feature_to_insert } =
+          create_target_objects(tree_object, tree_attributes);
+
+        //field_data.raw_capture insertion
+        await trx.table("field_data.raw_capture").insert(raw_capture_to_insert);
+
+        // map_feature.raw_capture_feature
+        const columns_to_insert = [
+          ...Object.keys(raw_capture_feature_to_insert),
+          "location",
+        ];
+        const { lon, lat } = raw_capture_feature_to_insert;
+        const values_to_insert = [
+          ...Object.values(raw_capture_feature_to_insert),
+        ];
+
+        await trx.raw(
+          `
+          INSERT INTO map_features.raw_capture_feature 
+          (
+            ${columns_to_insert.join(",")}
+          )
+          VALUES (${values_to_insert
```

Comment:
While this is clever use of objects, map functions, I would prefer to write simple queries. That way it is easier to read and understand.